### PR TITLE
Admin Mode Patch for launcher

### PR DESCRIPTION
**Pull request by dartasen**
Detect when launcher is running inside C:\Program Files and warn that the launcher likely needs admin permissions to properly install Nitrox into Subnautica. Provide an option in the popup to restart the launcher with admin permissions

### CODE REVIEW

File: NitroxLauncher/App.xaml.cs [hunk 1]
```diff
@@ -30,6 +30,31 @@ protected override void OnStartup(StartupEventArgs e)
             {
                 MessageBox.Show("Nitrox launcher should not be executed from a temporary directory. Install Nitrox launcher properly by extracting ALL files and moving these to a dedicated location on your PC.", "Invalid working directory", MessageBoxButton.OK, MessageBoxImage.Error);
                 Environment.Exit(1);
+                Shutdown();
```
Comment (Measurity) [inline NitroxLauncher/App.xaml.cs]: Environment.Exit will kill the app before this call so `Shutdown()` is unreachable.

Either remove the `Environment.Exit(1)` or remove `Shutdown()`.

File: NitroxLauncher/App.xaml.cs [hunk 2]
```diff
@@ -30,6 +30,31 @@ protected override void OnStartup(StartupEventArgs e)
...
+            }
+
+            if (!RoleDetection.isAppRunningInAdmin()) {
+
+                if (Directory.GetCurrentDirectory().StartsWith(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), StringComparison.OrdinalIgnoreCase))
```
Comment (Measurity) [inline NitroxLauncher/App.xaml.cs]: This should check for subnautica install path. Not the launcher working directory. If Subnautica is installed in Program Files then the launcher likely needs admin.

Comment (Coding-Hen) [inline NitroxLauncher/App.xaml.cs]: You should be able to just get the subnautica directory from the launcherlogic instance
```suggestion
                if (LauncherLogic.Instance.SubnauticaPath.StartsWith(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles), StringComparison.OrdinalIgnoreCase))
```

Comment (dartasen) [reply NitroxLauncher/App.xaml.cs]: He won't work, LauncherLogic isn't up yet while entering the application

**Comment by dartasen** [comment]
Well it's working as you requested to but as long as ProgramFiles is is most of the cases the directory used for playing Subnautica, will it not be better to always ask to have admin rights throught the app.manifest instead of often running "hacky code" to restart the launcher ?

File: NitroxLauncher/Patching/AppHelper.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        public static string ProgramFileDirectory = Environment.ExpandEnvironmentVariables("%ProgramW6432%");
+
+        public static bool IsAppRunningInAdmin()
+        {
+            WindowsPrincipal wp = new WindowsPrincipal(WindowsIdentity.GetCurrent());
+            return wp.IsInRole(WindowsBuiltInRole.Administrator);
+        }
+
+        public static void RestartAsAdmin(string path)
```
Comment (Measurity) [inline NitroxLauncher/Patching/AppHelper.cs]: Could you separate the logic from this method out into two?
`RestartAsAdmin` sounds good if it only restarted the app but it does more than that: checks the path of program files. 

You can remove the comment in the `SubnauticaPath` property if you change the code to something like:
```cs
if (!AppHelper.IsRunningAsAdmin() && Path.GetFullPath(path).StartsWith(ProgramFileDirectory, StringComparison.OrdinalIgnoreCase))
{
    AppHelper.RestartAsAdmin();
}
```

File: NitroxLauncher/LauncherLogic.cs [hunk 4]
```diff
@@ -102,6 +100,7 @@ public async Task<string> SetTargetedSubnauticaPath(string path)
                         nitroxEntryPatch.Remove();
                     }
                     nitroxEntryPatch = new NitroxEntryPatch(path);
+                    AppHelper.RestartAsAdmin(path); //If the change path to ProgramFiles, then we ask for admin perm
```
Comment (Measurity) [inline NitroxLauncher/LauncherLogic.cs]: Change this code so comment is not needed (see other comment).

**Pull request merged**